### PR TITLE
[sqlqueryreceiver] - add optional `static_attributes` to metrics. 

### PR DESCRIPTION
**Pull request by Konig-Corey**
**Description:** 

Adding a feature:
 - Add ability to set static tags for metrics returned from queries

**Testing:** 
- Added config test to validate config is loading a map of strings for tags. 

**Documentation:** 
- Entry added to README.md to specify static tag usage. 
FYI @pmcollins

### CODE REVIEW

File: receiver/sqlqueryreceiver/metrics.go [hunk 1]
```diff
@@ -27,6 +29,8 @@ func rowToMetric(row metricRow, cfg MetricCfg, dest pmetric.Metric) error {
 	dest.SetUnit(cfg.Unit)
 	dataPointSlice := setMetricFields(cfg, dest)
 	dataPoint := dataPointSlice.AppendEmpty()
+	dataPoint.SetStartTimestamp(pcommon.NewTimestampFromTime(time.Now()))
```
Comment (dmitryax) [inline receiver/sqlqueryreceiver/metrics.go:32]: start time cannot be set to the same timestamp. If has different meaning depending on the data type.
- Gauge: shouldn't have a start time.
- Cumulative sum should have a start time set to the beginning of the data points cumulation,
- Non-cumulative sum should have a start time set to the previous endpoint

Check https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/data-model.md#metric-points for more details

**Comment by dmitryax** [comment]
Also it would be nice to have this PR broken into several ones since the fixes are unrelated to each other. And don't forget to add change log entries :)

File: receiver/sqlqueryreceiver/metrics.go (hunk 1 above)
Comment (Konig-Corey) [reply receiver/sqlqueryreceiver/metrics.go:32]: Unless the receiver maintains the state of previous metrics how would one get the previous points timestamp OR the timestamp of the original metric? @dmitryax  

@pmcollins Any recommendations ?

**Comment by Konig-Corey** [comment]
I can close this and split it into to. I agree it makes more sense to have two PRs one for the addition of static tags, and another for the timestamp addition.

File: receiver/sqlqueryreceiver/metrics.go (hunk 1 above)
Comment (dmitryax) [reply receiver/sqlqueryreceiver/metrics.go:32]: Just `current time - scrape duration` should be sufficient for non-cumulative sum. And the receiver start time should be ok for cumulative sum.

Not sure why do we need sums at all to be honest. Gauge seems more appropriate for this kind of random data

**Comment by Konig-Corey** [comment]
@dmitryax removed the timestamp fix and will move to a different PR. Leaving this open for addition of static tag feature.

**Comment by dmitryax** [comment]
@Konig-Corey plase update PR description

File: receiver/sqlqueryreceiver/config.go [hunk 2]
```diff
@@ -80,6 +80,7 @@ type MetricCfg struct {
 	Aggregation      MetricAggregation `mapstructure:"aggregation"`
 	Unit             string            `mapstructure:"unit"`
 	Description      string            `mapstructure:"description"`
+	Tags             map[string]string `mapstructure:"tags"`
```
Comment (dmitryax) [inline receiver/sqlqueryreceiver/config.go:83]: Why do you think this functionality is essential part of the receiver? Addition of custom attributes can be easily achieved by using attributes processor https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/processor/attributesprocessor

Comment (Konig-Corey) [reply receiver/sqlqueryreceiver/config.go:83]: Using the processor requires intercepting the metric by name and adding the tags. Totally doable, but seems way more intuitive to just be able to set static tag values at the receiver config. In our case we have 20+ queries all generating a fair amount of metrics and having the ability to set the config at the receiver is more maintainable then setting it at the processor. Not to mention the overhead is next to nothing to have this config option on the receiver.

Comment (dmitryax) [reply receiver/sqlqueryreceiver/config.go:83]: Makes sense. Thanks for providing the reasons. 

Let's just rename the config option to OTel terminology. Maybe `extra_attributes`, `metric_attributes` or just `attributes`?

Comment (dmitryax) [reply receiver/sqlqueryreceiver/config.go:83]: or `static_attributes`

**Comment by dmitryax** [comment]
and a changelog entry is still required

File: CHANGELOG.md [hunk 3]
```diff
@@ -37,7 +37,6 @@
 - `sqlqueryreceiver`: Enable the component (#11848)
 
 ### 💡 Enhancements 💡
-
```
Comment (dmitryax) [inline CHANGELOG.md:40]: Unrelated removal :)

**Comment by dmitryax** [comment]
We just recently introduced another way to handle changelog. Please follow the updated process https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/CONTRIBUTING.md#adding-a-changelog-entry

**Comment by Konig-Corey** [comment]
@dmitryax Added new change log.

File: receiver/sqlqueryreceiver/README.md [hunk 4]
```diff
@@ -30,6 +30,7 @@ Each _metric_ in the configuration will produce one OTel metric per row returned
 * `aggregation` (optional): only applicable for `data_type=sum`; can be `cumulative` or `delta`; defaults to `cumulative`.
 * `description` (optional): the description applied to the metric.
 * `unit` (optional): the units applied to the metric.
+* `static_attributes` (optional): static labels applied to the metrics
```
Comment (dmitryax) [inline receiver/sqlqueryreceiver/README.md:33]:
```suggestion
* `static_attributes` (optional): static attributes applied to the metrics
```

File: unreleased/sqlqueryreceiver-tags.yaml [hunk 5]
```diff
@@ -0,0 +1,11 @@
...
+change_type: enhancement
+
+# The name of the component, or a single word describing the area of concern, (e.g. filelogreceiver)
+component: sqlqueryreceiver
+
+# A brief description of the change
+note: Add `static_attributes` optional configuration to allow adding attributes/tags to queried metrics. 
+
+# One or more tracking issues related to the change
+issues: []
```
[the remark this answers is not in the document]
Comment (Konig-Corey) [reply unreleased/sqlqueryreceiver-tags.yaml]: Added pr # to issues field 👍

**Comment by dmitryax** [comment]
@Konig-Corey please rebase

File: go.mod [hunk 6]
```diff
@@ -923,3 +923,11 @@ retract v0.37.0 // Contains dependencies on v0.36.0 components, which should hav
 
 // see https://github.com/distribution/distribution/issues/3590
 exclude github.com/docker/distribution v2.8.0+incompatible
+
+// Replacing prometheus/client_golang so that cmd/otelcontribcol/main_testbed.go passes. This is needed because the test runs exporter/prometheusexporter that requires the exemplars feature in github.com/prometheus/client_golang.
+// This replace can be removed once the exemplars feature (https://github.com/prometheus/client_golang/pull/986) is available in a release version and this go.mod is updated to use that.
+replace github.com/prometheus/client_golang => github.com/prometheus/client_golang v1.12.2-0.20220318110013-3bc8f2c651ff
```
Comment (dmitryax) [inline go.mod:929]: Why is this added? I believe this was reverted in https://github.com/open-telemetry/opentelemetry-collector-contrib/commit/282dbe70239e057f6f9246eb1e60e3bca9e8c828#diff-cbc1780ee44619ffeeb078f84991774b14445d3a192c34b6acc709f00788b421 and few tests passed after that

**Comment by dmitryax** [comment]
please take a look at CI failures

File: receiver/sqlqueryreceiver/README.md [hunk 7]
```diff
@@ -44,6 +45,8 @@ receivers:
           - metric_name: movie.genres
             value_column: "count"
             attribute_columns: [ "genre" ]
+            static_attributes: 
+               dbinstance: mydbinstance
```
Comment (mx-psi) [inline receiver/sqlqueryreceiver/README.md:49]: What if I want to set an attribute to `true`? How will we add this if we need it in a backwards compatible way?

File: receiver/sqlqueryreceiver/README.md [hunk 8]
```diff
@@ -30,6 +30,7 @@ Each _metric_ in the configuration will produce one OTel metric per row returned
 * `aggregation` (optional): only applicable for `data_type=sum`; can be `cumulative` or `delta`; defaults to `cumulative`.
 * `description` (optional): the description applied to the metric.
 * `unit` (optional): the units applied to the metric.
+* `static_attributes` (optional): static attributes applied to the metrics
```
Comment (mx-psi) [inline receiver/sqlqueryreceiver/README.md:33]: What is the benefit of using this option instead of just using the transform processor?

File: go.mod (hunk 6 above)
Comment (Konig-Corey) [reply go.mod:929]: No idea, must have missed it in the rebase.

File: receiver/sqlqueryreceiver/README.md (hunk 8 above)
Comment (Konig-Corey) [reply receiver/sqlqueryreceiver/README.md:33]: Please review answer above in thread.

File: receiver/sqlqueryreceiver/README.md (hunk 7 above)
Comment (Konig-Corey) [reply receiver/sqlqueryreceiver/README.md:49]: Im not sure what you are asking. It only takes string values. Any non string values would likely cause an issue.

File: go.mod (hunk 6 above)
Comment (Konig-Corey) [reply go.mod]: Rebased again to grab latest, and made sure this didn't get excluded. Got a little funky with all the conflicts/reverts.

File: receiver/sqlqueryreceiver/config.go (hunk 2 above)
Comment (Konig-Corey) [reply receiver/sqlqueryreceiver/config.go]: @mx-psi rational for config addition.

File: receiver/sqlqueryreceiver/README.md (hunk 7 above)
Comment (dmitryax) [reply receiver/sqlqueryreceiver/README.md:49]: The @mx-psi's point is that if we want to add attributes with different value types, we won't be able to do that without breaking existing configuration interface. So it's better to design it better in advance. Maybe something like:

```
static_attributes: 
  key: dbinstance
  string_value: mydbinstance
```

File: receiver/sqlqueryreceiver/README.md (hunk 8 above)
Comment (dmitryax) [reply receiver/sqlqueryreceiver/README.md:33]: Yeah, I asked the same Q here https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/11868#discussion_r912239391

**Comment by Konig-Corey** [comment]
> 

Working to fix it. Sorry about that.

File: receiver/sqlqueryreceiver/README.md (hunk 8 above)
Comment (mx-psi) [reply receiver/sqlqueryreceiver/README.md:33]: Thanks, that answers my question indeed!

File: receiver/sqlqueryreceiver/README.md (hunk 7 above)
Comment (dmitryax) [reply receiver/sqlqueryreceiver/README.md:49]: Are we still applying this configuration interface ^ ? @mx-psi what do you think bout it?

Comment (mx-psi) [reply receiver/sqlqueryreceiver/README.md:49]: I think changing `StaticAttributes` to `map[string]interface{}` and then doing:

```
for k, v := range cfg.StaticAttributes {
     switch v.(type) {
         case string:
	     attrs.InsertString(k, v)
          // ... other cases here, or even just a default that returns an error saying non-string is not supported
     }
```

Should work, and we don't have to complicate the configuration schema. Does that make sense?

Comment (dmitryax) [reply receiver/sqlqueryreceiver/README.md:49]: I'm not sure `configmap` allows that and yaml is not the best format for that, e.g. what value type do you expect from the following pair:
```yaml
static_attributes:
  bool_or_str: yes
```

Comment (mx-psi) [reply receiver/sqlqueryreceiver/README.md:49]: Our config provider definitely allows doing that: https://go.dev/play/p/CsyC3KuTer5

The yes/no case is an interesting edge case. The YAML package we use documents this explicitly [here](https://pkg.go.dev/gopkg.in/yaml.v3#readme-compatibility), in this case it will be a string since the field we unmarshal into is not a bool (see playground above). I think that's an acceptable behavior for me (but I understand if you feel otherwise).

Comment (dmitryax) [reply receiver/sqlqueryreceiver/README.md:49]: I'm ok with this approach. @Konig-Corey can you please keep existing interface but make sure that only string map values are excepted for now. Later we can add support for other types if needed

Comment (dmitryax) [reply receiver/sqlqueryreceiver/README.md:49]: After giving it a second thought, I would say I'd rather prefer explicit type definition in yaml config rather than relying on implying type from the value. As a user, I would rather be explicit about what type of values I will get, instead of figuring out what's the parsing logic to make sure that attribute has a type that I want. I don't think users will be setting many static attributes, usually one or two per metric, having another line in the settings per attribute is not a big overhead IMO. 

Specifying attributes in a list also makes the emitted metric with attributes in predictable order.

Comment (mx-psi) [reply receiver/sqlqueryreceiver/README.md:49]: I guess not everyone has the same familiarity with YAML's quirkiness so that's a good argument. 

Another reason I did not mention before to favor the map approach is that we sort-of have that on core [here](https://github.com/open-telemetry/opentelemetry-collector/blob/58e66af2558a56be1f26c759db80e0ed35234607/service/telemetry/config.go#L28-L32). I don't feel very strongly in favor of either implementation, but we should probably be consistent on both places. I don't want to drag this on forever, so I am fine with trying the explicit type here and then deciding which one we like the most.

Comment (dmitryax) [reply receiver/sqlqueryreceiver/README.md:49]: Makes sense, let's keep the same approach in this processor. I believe we can keep existing config behavior, given that the component is in alpha state. We can add restriction to string only attribute later.

**Review by dmitryax — APPROVED** [review]
LGTM. Sorry for the delay

**Comment by dmitryax** [comment]
@mx-psi PTAL
